keanu-project/src/main/java/io/improbable/keanu/tensor/INDArrayShim.java

```diff
@@ -99,6 +142,115 @@ private static INDArray broadcastMinus(INDArray a, INDArray b) {
+        return performOperationWithScalarTensorPreservingShape(left, right, Transforms::atan2);
+    }
+
+    public static INDArray lt(INDArray left, INDArray right) {
+        return performOperationWithScalarTensorPreservingShape(left, right, INDArray::lt);
+    }
+
+    public static INDArray gt(INDArray left, INDArray right) {
+        return performOperationWithScalarTensorPreservingShape(left, right, INDArray::gt);
+    }
+
+    public static INDArray eq(INDArray left, INDArray right) {
+        return performOperationWithScalarTensorPreservingShape(left, right, INDArray::eq);
+    }
+
+    private static INDArray performOperationWithScalarTensorPreservingShape(INDArray left, INDArray right, BiFunction<INDArray, INDArray, INDArray> operation) {
+        if (left.length() == 1 || right.length() == 1) {
+            long[] resultShape = Shape.broadcastOutputShape(left.shape(), right.shape());
+            INDArray result;
+            if (left.length() == 1) {
+                result = operation.apply(Nd4j.valueArrayOf(right.shape(), left.getDouble(0)), right);
+            } else {
+                result = operation.apply(left, Nd4j.valueArrayOf(left.shape(), right.getDouble(0)));
+            }
+            return result.reshape(resultShape);
+        }
+        return operation.apply(left, right);
+    }
+
+    @FunctionalInterface
+    interface QuadFunction<First, Second, Third, Fourth, Result> {
+        Result apply(First one, Second two, Third three, Fourth four);
+    }
+
+    public static INDArray executeNd4jTransformOpWithPreservedScalarTensorShape(INDArray mask, INDArray right, DataBuffer.Type bufferType, QuadFunction<INDArray, INDArray, INDArray, Long, BaseTransformOp> baseTransformOpConstructor) {
+        if (mask.length() == 1 || right.length() == 1) {
+            long[] resultShape = Shape.broadcastOutputShape(mask.shape(), right.shape());
+            if (mask.length() == 1) {
+                mask = Nd4j.valueArrayOf(right.shape(), mask.getDouble(0));
+                Nd4j.getExecutioner().exec(
+                    baseTransformOpConstructor.apply(mask, right, mask, mask.length())
+                );
+            } else {
+                Nd4j.getExecutioner().exec(
+                    baseTransformOpConstructor.apply(mask,
+                        valueArrayOf(mask.shape(), right.getDouble(0), bufferType),
+                        mask,
+                        mask.length()
+                    )
+                );
+            }
+            return mask.reshape(resultShape);
+        }
+        Nd4j.getExecutioner().exec(
+            baseTransformOpConstructor.apply(mask, right, mask, mask.length())
+        );
+        return mask;
+    }
+
+    public static INDArray getGreaterThanMask(INDArray mask, INDArray right, DataBuffer.Type bufferType) {
+        QuadFunction<INDArray, INDArray, INDArray, Long, BaseTransformOp> oldGreaterThanConstructor = OldGreaterThan::new;
```

Comment:
this variable can be inlined. 

ditto for the other methods that call `executeNd4jTransformOpWithPreservedScalarTensorShape`

Reply:
done